pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py

```diff
@@ -28,6 +27,40 @@ def entry_point():
     return PuppetModuleInstallDistributor, {}
 
 
+class NormalizingTarInfo(tarfile.TarInfo):
+    """
+    Use this class with a call to tarfile.open(). It ensures that the uid and gid of extracted files
+    get set to the values of the current process, and that the permissions on files and dirs get
+    the below minimum values.
+
+    Specifically, it is important to guarantee that puppet can read the extracted files.
+    """
+    min_file_perms = int('0644', base=8)
+    min_dir_perms = int('0755', base=8)
+
+    @classmethod
+    def fromtarfile(cls, *args, **kwargs):
+        """
+        This removes the uid and gid from the returned TarInfo file, causing those values to be set
+        based on the process currently running.
+
+        This also adds a basic level of filesystem permissions to each file and directory extracted,
+        to ensure that they can be accessed and moved around as required.
+
+        :return:    a new TarInfo instance representing a file in an archive
+        :rtype:     tarfile.TarInfo
+        """
+        ret = super(NormalizingTarInfo, cls).fromtarfile(*args, **kwargs)
+        # this causes the uid and gid to be set matching the process doing the extraction
+        ret.uid = 0
+        ret.gid = 0
```

Comment:
If the worker process is run by root, the default behavior is to attempt to write files with the uid and gid from the tarball. I think that is worth guarding against, and this is a very simple behavior that is not likely to cause confusion.

The docs also have a vague implication that other platforms may allow "normal" users to assign a different uid or gid. There are no additional details, and this is admittedly unlikely even for an edge case, but we can guard against that also with these very simple 2 lines of code.